chainermn/dataset.py

```diff
@@ -24,24 +27,33 @@ def scatter_dataset(dataset, comm):
         comm = comm.mpi_comm
     assert hasattr(comm, 'send')
     assert hasattr(comm, 'recv')
+    assert 0 <= root and root < comm.size
 
     # We cannot use `mpi_comm.scatter`. This is due to MPI4py's bug.
     # For large datasets, when using `mpi_comm.scatter`, it causes MemoryError.
-    if comm.rank == 0:
+    # import sys
+    # sys.stderr.write("scatter_dataset(): root={}".format(root))
+    if comm.rank == root:
         mine = None
         n_total_samples = len(dataset)
         n_sub_samples = (n_total_samples + comm.size - 1) // comm.size
+
+        if shuffle:
+            order = numpy.random.RandomState(seed).permutation(n_total_samples)
+        else:
+            order = numpy.arange(n_total_samples)
```

Comment:
`order = None` will also work, and it would probably bring (slight) memory/network usage improvement.